text/0006-sampling.md

```diff
@@ -0,0 +1,300 @@
+ * Allow application operator to configure the sampling frequency.
+ 
+For new modern systems that need to collect all the Spans and later may or may not do a deferred 
+sampling decision, the OpenTelemetry needs to natively support a way to configure the library to 
+collect and export all the Spans. This is possible even though OpenTelemetry supports sampling by
+setting a default config to always collect all the spans.
+
+### Sampling flags
+OpenTelemetry API has two flags/properties:
+ * `RecordEvents`
+   * This property is exposed in the `Span` interface.
+   * If `true` the current `Span` records tracing events (attributes, events, status, etc.), 
+   otherwise all tracing events are dropped.
+   * Users can use this property to determine if expensive trace events can be avoided.
+ * `SampledFlag` 
+   * This flag is propagated via the `TraceOptions` (may be renamed to `TraceFlags` in a different
+   PR) to the child Spans. For more details see the w3c definition [here][trace-flags].
+   * In Dapper based systems this is equivalent to `Span` being `sampled` and exported.
+   
+The flag combination `SampledFlag == false` and `RecordEvents == true` means that the current `Span`
+does record tracing events, but most likely the child `Span` will not. This combination is 
+necessary because:
+ * Allow users to control recording for individual Spans.
+ * OpenCensus has this to support z-pages, so we need to keep backwards compatibility.
+
+The flag combination `SampledFlag == true` and `RecordEvents == false` can cause gaps in the 
+distributed trace, and because of this OpenTelemetry API should NOT allow this combination.
+
+It is safe to assume that users of the API should only access the `RecordEvents` property when 
+instrumenting code and never access `SampledFlag` unless used in context propagators.
+
+### SamplingHint
+This is a new concept added in the OpenTelemetry API that allows to suggest sampling hints to the
+implementation of the API:
+ * `UNSPECIFIED`
+   * This is the default option.
+   * No suggestion is made.
+ * `NOT_RECORD`
+   * Suggest to not `RecordEvents = false` and not propagate `SampledFlag = false`.
+ * `RECORD`
+   * Suggest `RecordEvents = true` and `SampledFlag = false`.
+ * `RECORD_AND_PROPAGATE`
+   * Suggest to `RecordEvents = true` and propagate `SampledFlag = true`.
+
+### Sampler interface
+The interface for the Sampler class that is available only in the OpenTelemetry SDK:
+ * `TraceID`
+ * `SpanID`
+ * Parent `SpanContext` if any
+ * `SamplerHint`
+ * `Links`
+ * Initial set of `Attributes` for the `Span` being constructed
+
+It produces as an output called `SamplingResult`:
+ * A `SamplingDecision` enum [`NOT_RECORD`, `RECORD`, `RECORD_AND_PROPAGATE`].
+ * A set of span Attributes that will also be added to the `Span`.
+   * These attributes will be added after the initial set of `Attributes`.
+ * (under discussion in separate RFC) the SamplingRate float.
```

Comment:
How about span name?

Reply:
I don't think the Sampler should change the Span name.

Reply:
Apologies I commented on the wrong line - I mean the `Sample` interface may need to accept span names as input parameter in order to allow sampling on particular namespaces. /cc @rghetia brought this up before.

Reply:
Good catch, I forgot about that.

Done.